### PR TITLE
[Ruby 3.0] Bump rubocop version

### PR DESCRIPTION
**Pull request by ainame**
This PR has about 600 LoC changes in automatically generated files (`.rubocop_todo.yml` and `Gemfile.lock`) so the actual changes are small. Don't be scared😇 

### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Prior to this PR, as per #17931, I've been working on Ruby 3.0 support tasks. This PR bumps the long-standing very old rubocop version (0.49.1) to the latest (1.12.1) and hopefully enables us to keep it up-to-date. rubocop is the most popular linter tool in Ruby and essential for fastlane to maintain the codebase consistent. However, the version we fix doesn't support Ruby 3.0 due to the version requirement spec in Rubocop. This gem https://github.com/whitequark/parser to parse Ruby code, whose rubocop depends on, don't support Ruby 3.0 until 3.0.0.0. Meaning we won't be able to run the linter against Ruby 3.0 when it gets supported in fastlane until bumping the rubocop version💥 This is how gemspec in rubocop v0.49.1 looks like 😇 
https://github.com/rubocop/rubocop/blob/1c2245df09cf0a1659b83acf0abb473e9b927a1a/rubocop.gemspec#L29

That's why this PR is here🙂 

### Description

The thing is that we need to catch up with many changes in rubocop. rubocop is adding new rules(cops) and improving existing ones continuously so that we can see a lot of warnings when running the latest version against fastlane. I'm going to introduce `.rubocop_todo.yml` to address that problem. 

`.rubocop_todo.yml` is an automatically generated file that is meant to disable overwhelming rubocop's warnings for existing project _for now_ in order to fix them gradually. This way we don't really need to fix such issues pointed by rubocop in this bumping PR and we can improve them later in future PRs.

https://docs.rubocop.org/rubocop/configuration.html#automatically-generated-configuration

Besides that what I did in this PR are -

* Fixed `.rubocop.yml` to be able to run the latest rubocop (there are some cops that aren't supported any more)
* Migrated some cops's department (namespace) as those are renamed to the different department; i.e. Style/VariableName -> Naming/VariableName
* Added `rubocop-performance` to gemspec as it spun off from rubocop gem
* Fixed custom cops implemented in fastlane to catch up on the latest API in rubocop 
* Fixed some issues pointed that are missed by rubocop even though we enabled the rules 6a0598f4145e1a6926d8fdc836418e102dee735b
* Fixed how the plugin's .rubocop.yml is generated to make it work
* Update one line of code to stop rubocop's crash https://github.com/fastlane/fastlane/pull/18564/commits/14ea64ad5f13b7211fb805db0d5a395e1fa036e3

I will add more information about changes with inline comments.

---

Next, I will submit another PR that applies auto-correction to fastlane in one go based on the latest rubocop so that we can fix many issues in TODO quickly. And then anyone will be able to contribute by picking issues in `.rubocop_todo.yml`.

Note that rubocop 1.13.0 was released on 20th April but it dropped Ruby 2.4 support so we can't upgrade to it for now but once we support Ruby 3.0 we can keep rubocop up-to-date again🙂  

### Testing Steps

See if CI passes linter checks

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -1,9 +1,28 @@
+inherit_from: .rubocop_todo.yml
+
 require:
   - rubocop/require_tools
+  - rubocop-performance
   - ./rubocop/fork_usage.rb
   - ./rubocop/missing_keys_on_shared_area.rb
   - ./rubocop/is_string_usage.rb
 
+AllCops:
```
Comment (ainame) [inline .rubocop.yml:10]: This is moved to the top to make it prominent.

File: .rubocop.yml [hunk 2]
```diff
@@ -1,9 +1,28 @@
...
+  TargetRubyVersion: 2.4
+  NewCops: enable
+  Include:
+    - '**/*.rb'
+    - '**/*file'
+    - '**/*.gemspec'
+    - '*/lib/assets/*Template'
+    - '*/lib/assets/*TemplateAndroid'
```
Comment (ainame) [inline .rubocop.yml:18]: This is now required to declare what is included explicitly.

File: .rubocop_todo.yml [hunk 3]
```diff
@@ -0,0 +1,575 @@
+# This configuration was generated by
+# `rubocop --auto-gen-config --exclude-limit 0`
+# on 2021-04-21 06:50:09 UTC using RuboCop version 1.12.1.
+# The point is for the user to remove these configuration records
+# one by one as the offenses are removed from the code base.
+# Note that changes in the inspected code, or installation of new
+# versions of RuboCop, may require this file to be generated again.
```
Comment (ainame) [inline .rubocop_todo.yml:7]: This file is generated automatically. So no need to review unless you are interested in seeing what rules are added😄

File: .rubocop.yml [hunk 4]
```diff
@@ -191,16 +207,13 @@ Metrics/ClassLength:
...
-Metrics/LineLength:
+Layout/LineLength:
   Max: 370
 
 # Configuration parameters: CountKeywordArgs.
 Metrics/ParameterLists:
   Max: 17
 
-Metrics/PerceivedComplexity:
-  Max: 18
```
Comment (ainame) [inline .rubocop.yml:202]: It seems that logic was changed and some files don't meet this threshold any more. So I removed it here and rubocop put "Max: 26" in the TODO file.

File: fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb [hunk 5]
```diff
@@ -1,6 +1,4 @@
-# coding: utf-8
-
-lib = File.expand_path("../lib", __FILE__)
+lib = File.expand_path("lib", __dir__)
```
Comment (ainame) [inline fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb:1]: Changes in this file are pointed by rubocop. Unlike others, this is needed to be fixed as issues will be passed to plugins newly generated😇

File: fastlane_core/lib/fastlane_core/languages.rb [hunk 6]
```diff
@@ -3,14 +3,14 @@ module Languages
     # These are all the languages which are available to use to upload app metadata and screenshots
 
     # The old format which was used until August 2015 (good old times)
-    ALL_LANGUAGES_LEGACY = ["da-DK", "de-DE", "el-GR", "en-AU", "en-CA", "en-GB", "en-US", "es-ES", "es-MX", "fi-FI", "fr-CA", "fr-FR", "id-ID", "it-IT", "ja-JP", "ko-KR", "ms-MY", "nl-NL", "no-NO", "pt-BR", "pt-PT", "ru-RU", "sv-SE", "th-TH", "tr-TR", "vi-VI", "cmn-Hans", "cmn-Hant"]
+    ALL_LANGUAGES_LEGACY = %w[da-DK de-DE el-GR en-AU en-CA en-GB en-US es-ES es-MX fi-FI fr-CA fr-FR id-ID it-IT ja-JP ko-KR ms-MY nl-NL no-NO pt-BR pt-PT ru-RU sv-SE th-TH tr-TR vi-VI cmn-Hans cmn-Hant]
```
Comment (ainame) [inline fastlane_core/lib/fastlane_core/languages.rb:6]: This exceeded the limit but for some reasons, it wasn't detected.

```yaml
# The %w might be confusing for new users
Style/WordArray:
  MinSize: 19
```

File: rubocop/fork_usage.rb [hunk 7]
```diff
@@ -33,7 +33,7 @@ def on_if(node)
       def on_send(node)
         return unless bad_fork(node)
         return if self.good_nodes.include?(node)
-        add_offense(node, :expression, MSG)
+        add_offense(node, location: :expression, message: MSG)
```
Comment (ainame) [inline rubocop/fork_usage.rb:36]: `add_offsense` in rubocop no longer support positional arguments and instead, we need to use keyword arguments.

File: deliver/lib/deliver/languages.rb [hunk 8]
```diff
@@ -2,6 +2,6 @@ module Deliver
   module Languages
     # 2020-08-24 - Available locales are not available as an endpoint in App Store Connect
     # Update with Spaceship::Tunes.client.available_languages.sort (as long as endpoint is avilable)
-    ALL_LANGUAGES = ["ar-SA", "ca", "cs", "da", "de-DE", "el", "en-AU", "en-CA", "en-GB", "en-US", "es-ES", "es-MX", "fi", "fr-CA", "fr-FR", "he", "hi", "hr", "hu", "id", "it", "ja", "ko", "ms", "nl-NL", "no", "pl", "pt-BR", "pt-PT", "ro", "ru", "sk", "sv", "th", "tr", "uk", "vi", "zh-Hans", "zh-Hant"]
+    ALL_LANGUAGES = %w[ar-SA ca cs da de-DE el en-AU en-CA en-GB en-US es-ES es-MX fi fr-CA fr-FR he hi hr hu id it ja ko ms nl-NL no pl pt-BR pt-PT ro ru sk sv th tr uk vi zh-Hans zh-Hant]
```
Comment (ainame) [inline deliver/lib/deliver/languages.rb:5]: This exceeded the limit but for some reasons, it wasn't detected.

```yaml
# The %w might be confusing for new users
Style/WordArray:
  MinSize: 19
```

File: fastlane/lib/fastlane/cli_tools_distributor.rb [hunk 9]
```diff
@@ -23,7 +23,7 @@ def utf8_locale?
       def take_off
         before_import_time = Time.now
 
-        if !ENV["FASTLANE_DISABLE_ANIMATION"]
+        if ENV["FASTLANE_DISABLE_ANIMATION"].nil?
```
Comment (ainame) [inline fastlane/lib/fastlane/cli_tools_distributor.rb:26]: This fixes a crash in rubocop. I would report it to them but in the meantime, we can keep it running with this easy fix.

https://gist.github.com/ainame/b90ba9fc159a0ff2ab8cb417ac9586c0

Comment (minuscorp) [reply fastlane/lib/fastlane/cli_tools_distributor.rb:26]: How weird

File: fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb (hunk 5 above)
Comment (minuscorp) [reply fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb:1]: This will need an update of all plugins?

**Review by minuscorp — APPROVED** [review]
What a work here! I'm glad that you didn't take all cops by yourself, you did more than enough! ❤️🚀

**Review by joshdholtz — APPROVED** [review]
Oh my goodness... I can’t thank you enough for this PR 🙏  This was very much needed and you killed it. This is going to great ❤️

File: fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb (hunk 5 above)
Comment (ainame) [reply fastlane/lib/fastlane/plugins/template/%gem_name%.gemspec.erb:1]: No. This is just style fix. If plugin authors think they want to follow it, they can just do it by themselves.

**Pull request merged**